### Armut - Association Rule Based Recommender System

**İş Problemi**

Türkiye’nin en büyük online hizmet platformu olan Armut, hizmet verenler ile hizmet almak isteyenleri buluşturmaktadır.
Bilgisayar veya akıllı telefon üzerinden birkaç dokunuşla temizlik, tadilat, nakliyat gibi hizmetlere kolayca ulaşılmasını sağlamaktadır. 

Hizmet alan kullanıcıları ve bu kullanıcıların almış oldukları servis ve kategorileri içeren veri setini kullanarak Association
Rule Learning ile ürün tavsiye sistemi oluşturulmak istenmektedir.

**Veri Seti Hikayesi**

Veri seti müşterilerin aldıkları servislerden ve bu servislerin kategorilerinden oluşmaktadır. Alınan her hizmetin tarih ve saat
bilgisini içermektedir.

**Değişkenler**

**UserId:** Müşteri numarası

**ServiceId:** Her kategoriye ait anonimleştirilmiş servislerdir. (Örnek : Temizlik kategorisi altında koltuk yıkama servisi) 
Bir ServiceId farklı kategoriler altında bulanabilir ve farklı kategoriler altında farklı servisleri ifade eder. (Örnek: CategoryId’si 7 ServiceId’si 4 olan hizmet petek temizliği iken CategoryId’si 2 ServiceId’si 4 olan hizmet mobilya montaj)

**CategoryId:** Anonimleştirilmiş kategorilerdir. (Örnek : Temizlik, nakliyat, tadilat kategorisi)

**CreateDate:** Hizmetin satın alındığı tarih

#### Görev 1: Veriyi Hazırlama

**Adım 1:** armut_data.csv dosyasını okutunuz.

In [1]:
# Gerekli kütüphane ve fonksiyonların yüklenmesi

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

# !pip install mlxtend
from mlxtend.frequent_patterns import apriori, association_rules

armut_data = pd.read_csv("datasets/armut_data.csv")

df = armut_data.copy()

df.head()

UserId  ServiceId  CategoryId           CreateDate
0   25446          4           5  2017-08-06 16:11:00
1   22948         48           5  2017-08-06 16:12:00
2   10618          0           8  2017-08-06 16:13:00
3    7256          9           4  2017-08-06 16:14:00
4   25446         48           5  2017-08-06 16:16:00

In [2]:
df.shape

(162523, 4)

**Adım 2:** ServisID her bir CategoryID özelinde farklı bir hizmeti temsil etmektedir. ServiceID ve CategoryID’yi "_" ile birleştirerek bu hizmetleri
temsil edecek yeni bir değişken oluşturunuz. 

In [3]:
df["Hizmet"] = [str(row[1]) + "_" + str(row[2]) for row in df.values]

df.head()

UserId  ServiceId  CategoryId           CreateDate Hizmet
0   25446          4           5  2017-08-06 16:11:00    4_5
1   22948         48           5  2017-08-06 16:12:00   48_5
2   10618          0           8  2017-08-06 16:13:00    0_8
3    7256          9           4  2017-08-06 16:14:00    9_4
4   25446         48           5  2017-08-06 16:16:00   48_5

**Adım 3:** Veri seti hizmetlerin alındığı tarih ve saatten oluşmaktadır, herhangi bir sepet tanımı (fatura vb. ) bulunmamaktadır. Association Rule Learning uygulayabilmek için bir sepet (fatura vb.) tanımı oluşturulması gerekmektedir. **Burada sepet tanımı her bir müşterinin aylık aldığı hizmetlerdir.** Örneğin; 7256 id'li müşteri 2017'in 8.ayında aldığı 9_4, 46_4 hizmetleri bir sepeti; 2017’in 10.ayında aldığı 9_4, 38_4 hizmetleri başka bir sepeti ifade etmektedir. Sepetleri unique bir ID ile tanımlanması gerekmektedir. Bunun için öncelikle sadece yıl ve ay içeren yeni bir date değişkeni oluşturunuz. UserID ve yeni oluşturduğunuz date değişkenini kullanıcı bazında "_" ile birleştirirek ID adında yeni bir değişkene atayınız.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162523 entries, 0 to 162522
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   UserId      162523 non-null  int64 
 1   ServiceId   162523 non-null  int64 
 2   CategoryId  162523 non-null  int64 
 3   CreateDate  162523 non-null  object
 4   Hizmet      162523 non-null  object
dtypes: int64(3), object(2)
memory usage: 6.2+ MB


In [5]:
df["CreateDate"] = pd.to_datetime(df["CreateDate"])

df["New_Date"] = df["CreateDate"].dt.strftime("%Y-%m")

df.head()

UserId  ServiceId  CategoryId          CreateDate Hizmet New_Date
0   25446          4           5 2017-08-06 16:11:00    4_5  2017-08
1   22948         48           5 2017-08-06 16:12:00   48_5  2017-08
2   10618          0           8 2017-08-06 16:13:00    0_8  2017-08
3    7256          9           4 2017-08-06 16:14:00    9_4  2017-08
4   25446         48           5 2017-08-06 16:16:00   48_5  2017-08

In [6]:
df["SepetID"] = [str(row[0]) + "_" + str(row[5]) for row in df.values]

df.head()

UserId  ServiceId  CategoryId          CreateDate Hizmet New_Date  \
0   25446          4           5 2017-08-06 16:11:00    4_5  2017-08   
1   22948         48           5 2017-08-06 16:12:00   48_5  2017-08   
2   10618          0           8 2017-08-06 16:13:00    0_8  2017-08   
3    7256          9           4 2017-08-06 16:14:00    9_4  2017-08   
4   25446         48           5 2017-08-06 16:16:00   48_5  2017-08   

         SepetID  
0  25446_2017-08  
1  22948_2017-08  
2  10618_2017-08  
3   7256_2017-08  
4  25446_2017-08

#### Görev 2: Birliktelik Kuralları Üretiniz ve Öneride bulununuz.

**Adım 1:** Aşağıdaki gibi sepet, hizmet pivot table’i oluşturunuz.

In [7]:
invoice_product_df = df.groupby(["SepetID", "Hizmet"])["Hizmet"].count().unstack().fillna(0).applymap(lambda x: 1 if x > 0 else 0)

invoice_product_df.head()

Hizmet         0_8  10_9  11_11  12_7  13_11  14_7  15_1  16_8  17_5  18_4  \
SepetID                                                                      
0_2017-08        0     0      0     0      0     0     0     0     0     0   
0_2017-09        0     0      0     0      0     0     0     0     0     0   
0_2018-01        0     0      0     0      0     0     0     0     0     0   
0_2018-04        0     0      0     0      0     1     0     0     0     0   
10000_2017-08    0     0      0     0      0     0     0     0     0     0   

Hizmet         ...  46_4  47_7  48_5  49_1  4_5  5_11  6_7  7_3  8_5  9_4  
SepetID        ...                                                         
0_2017-08      ...     1     0     1     0    0     0    0    0    0    0  
0_2017-09      ...     0     0     1     0    1     0    0    0    0    0  
0_2018-01      ...     0     0     0     0    0     0    0    1    0    0  
0_2018-04      ...     0     0     0     0    0     0    0    0    0    0  
10000_2017-08  ...     1     0     0     0    0     0    0    0    0    0  

[5 rows x 50 columns]

**Adım 2:** Birliktelik kurallarını oluşturunuz.

In [8]:
frequent_itemsets = apriori(invoice_product_df, min_support = 0.01, use_colnames = True)

rules = association_rules(frequent_itemsets, metric = "support", min_threshold = 0.01)

rules.head()

C:\Users\ogusan\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents consequents  antecedent support  consequent support   support  \
0       (2_0)     (13_11)            0.130286            0.056627  0.012819   
1     (13_11)       (2_0)            0.056627            0.130286  0.012819   
2       (2_0)      (15_1)            0.130286            0.120963  0.033951   
3      (15_1)       (2_0)            0.120963            0.130286  0.033951   
4      (15_1)      (33_4)            0.120963            0.027310  0.011233   

   confidence      lift  leverage  conviction  
0    0.098394  1.737574  0.005442    1.046325  
1    0.226382  1.737574  0.005442    1.124216  
2    0.260588  2.154278  0.018191    1.188833  
3    0.280673  2.154278  0.018191    1.209066  
4    0.092861  3.400299  0.007929    1.072262

**Adım 3:** arl_recommender fonksiyonunu kullanarak son 1 ay içerisinde en son 2_0 hizmetini alan bir kullanıcıya hizmet önerisinde bulununuz.

In [9]:
def arl_recommender(rules_df, product_id, rec_count = 1):
    sorted_rules = rules_df.sort_values("lift", ascending = False)
    recommendation_list = []
    for i, product in enumerate(sorted_rules["antecedents"]):
        for j in list(product):
            if j == product_id:
                recommendation_list.append(list(sorted_rules.iloc[i]["consequents"])[0])
    return recommendation_list[0:rec_count]

In [10]:
arl_recommender(rules, "2_0,", 1)

[]

In [11]:
arl_recommender(rules, "2_0,", 2)

[]

In [12]:
arl_recommender(rules,"2_0", 3)

['22_0', '25_0', '15_1']

In [13]:
arl_recommender(rules,"2_0", 4)

['22_0', '25_0', '15_1', '13_11']